module/zfs/dsl_scan.c

```diff
@@ -403,6 +423,56 @@ dsl_scan_cancel(dsl_pool_t *dp)
 	    dsl_scan_cancel_sync, NULL, 3, ZFS_SPACE_CHECK_RESERVED));
 }
 
+/*
+ * Set scrub pause/resume state if it makes sense to do so
+ */
+int
+dsl_scrub_set_pause_resume(const dsl_pool_t *dp, pool_scrub_cmd_t cmd)
+{
+	spa_t *spa = dp->dp_spa;
+	dsl_scan_t *scn = dp->dp_scan;
+
+	ASSERT(cmd == POOL_SCRUB_NORMAL || cmd == POOL_SCRUB_PAUSE);
+
+	/* can't pause a scrub when there is no in-progress scrub */
+	if (cmd == POOL_SCRUB_PAUSE && !dsl_scan_scrubbing(dp))
+		return (SET_ERROR(ENOENT));
+
+	/* can't pause a paused scrub */
+	if (cmd == POOL_SCRUB_PAUSE && dsl_scan_is_paused_scrub(scn))
+		return (SET_ERROR(EBUSY));
+
+	if (cmd == POOL_SCRUB_PAUSE) {
+		spa->spa_scan_pass_scrub_pause = gethrestime_sec();
+		scn->scn_phys.scn_flags |= DSF_SCRUB_PAUSED;
```

Comment:
Is it OK to do this from open context (I'm guessing no)?  When will it be persisted to disk (I'm guessing never / by luck, since you are not calling dsl_scan_state_sync(), which can't be called from open context anyway)?

I would suggest that you do this from a synctask instead.